### ISSUE TITLE
Re-implement 'fullres' argument, to allow user-specified fullscreen resolution

### ISSUE THREAD

**Issue by sa666666** [stella-emu/stella#285]
A good use-case was recently pointed out on AtariAge:
http://atariage.com/forums/topic/274560-question-about-missing-50hz-fullscreenmode-in-newest-stella-version

I never considered this before, since I don't use PAL systems, and don't use a monitor with variable refresh rate.

**Comment by thrust26** [comment]
Looks like the option -fullres got lost with SDL2 (4.0).

An intermediate doc names the option "FIXSDL-fullres". So it sound like there were problems during the migration.

**Comment by DirtyHairy** [comment]
Please don't.

* VCS games don't display at fixed rates; we all know that the frame rate differs from game to game and sometimes even in-game
* Naively locking the frame rate to the external display will break cycle exact audio unless the sample rate is continuously adapted to the current FPS, which will decrease audio accuracy 
* CRT monitors are a dying species. Already now, almost everyone uses LCD panels which can't vary their refresh rate.
* I am 99.9% sure that this guy is chasing butterflies --- the same as audiophiles that swear on golden fuses.

The proper way to deal with beating effects between the display refresh rate and the frame stream generated by the TIA is properly sampling the time series of frames.

**Comment by thrust26** [comment]
I also don't think the frame rate should be locked. But the closer the monitor's frame rate matches the game's frame rate, the less frame drops/duplicates we have. And then e.g. scrolling looks smoother.

IMO the op only wants to have an option to adapt the monitor frame rate, not vice versa.

**Comment by sa666666** [comment]
TO be clear, I plan to only introduce the resolution to switch to in fullscreen mode, with the default to just use the desktop resolution (which is what happens all the time now).  It is simply a drop-down list of resolutions that SDL advertises it supports.  I wouldn't be implementing anything beyond that, since that was all was there pre-4.0.

I suppose it *does* serve a purpose to have it.  Some people run their games at different resolutions, to account for slow video cards, etc.  So it doesn't hurt to offer the option.

**Comment by thrust26** [comment]
Yes, that makes sense. I'll lookup that SDL resolution list.

**Comment by DirtyHairy** [comment]
I still wouldn't spend much time on this. This kind of tweaking caters more towards the placebo effect than anything else, and I doubt there is any machine out there that can handle Stella's emulation core and has trouble issues blitting the result to the screen ;)

Also, to be clear, there will be the necessity for changes to the way vsync and the display driver for the new audio code --- blocking will mess up the timing. If frame drops or dupes turn our to be a problem, I think one possible solution is to run the emulation one or two frame ahead and interpolate and sample between on screen refresh.

**Comment by thrust26** [comment]
We will have to support VSYNC anyway, else we will have tearing effects.

**Comment by DirtyHairy** [comment]
Definitely, but we cannot block anymore while waiting for vsync.

**Comment by DirtyHairy** [comment]
Ups 😄

**Comment by thrust26** [comment]
Happened to me too. And more than once! :smile:

**Comment by sa666666** [comment]
The whole point of having sync to vblank (aka v-sync) in the first place was to lock to the scan rate and eliminate tearing.  I would be interested to see how we can one (no tearing) without the other (no blocking).  Put another way, I thought that v-sync always implies blocking.

**Comment by DirtyHairy** [comment]
Cycle exact audio means that every sample corresponds to a fixed number of cycles, and this completely fixes the timing of the emulation loop. Blocking for screen refresh will mess up the timing and cause audio drops. The only way around would be dynamically adapting the sample rate to the actual timing, and doing this reliably while maintaining a steady audio stream seems close to impossible to me.

The approach that I have in mind is completely decoupling the emulation loop from the video driver:

* The emulation loop generates a steady stream of frames
* The video driver runs on a separate thread and always renders the last produced frame, synchronized with vsync.

This is what I do with 6502.ts, and it works well enough. The potential problem in this approach are frame drops or dupes that happen every now and them, but I have yet to see a clear indication that this is even visible. In case it is, the effect could be reduced by not simply using the last generated frame, but instead interpolating smoothly between the two closest frames in the time series.

**Comment by sa666666** [comment]
Everything you said makes sense; I just hope we don't lose smooth video output.  But I guess that's becoming less of a concern all the time, since most OS's now have composited desktops, where the desktop itself is double-buffered and v-sync'ed.

That's the way AMD cards work in Linux, at least.  Even if I turn vsync off in Stella, it still gives smooth updates because the sync is done at the OS level.  I believe Windows (10 at least, perhaps 7) and OSX also do this by default.

**Comment by sa666666** [comment]
This is post-6.0, since it needs more discussion.

**Comment by sa666666** [comment]
Just going over this again.  Should I add a dropdown box that allows to select the fullscreen video mode?  Right now, Stella just uses the desktop resolution as determined by SDL.  To be clear, I wouldn't be doing anything else, like looking at refresh rates, etc.  In fact, I don't even think SDL currently allows this.

This will simply be a setting to be able to specify a fullscreen resolution to use other than the default (desktop) resolution.  Is there any use-case for this?  I don't think it's too hard to add it; just wondering if it's worth it??

**Comment by thrust26** [comment]
I don't care much for the resolution (2600 pixels will never be square anyway), but different refresh rates make sense. E.g. 50Hz for PAL games makes sense. Maybe we should lock the resolution and only allow switching refresh rates? Maybe even automatically?

Actually I added some test code just yesterday to check what my system supports (only 60 and 75Hz).  :)

**Comment by sa666666** [comment]
I don't think SDL allows to select modes by refresh rate; I don't think it even exposes this information.  So all that this will do is allow the user to pick a resolution that may happen to have 50Hz.  And on any LCD from the past 10 years or so, that's probably 60Hz (or more, for newer ones).  I don't know of any LCD that does 50Hz, but I assume CRT will.  But again, you would have to pick different modes until one happens to have 50Hz.

**Comment by thrust26** [comment]
SDL_GetDisplayMode returns resolution, **refresh rate** and color information.

**Comment by sa666666** [comment]
Hmm, this must be something fairly new.  Well in that case, my original questions stands; is this worthwhile to add??

**Comment by thrust26** [comment]
If it is not too complex, why not?

**Comment by sa666666** [comment]
OK, I'll look into it tomorrow (leaving work soon, and taking this evening off).

**Comment by DirtyHairy** [comment]
While I know that SDL2 exposes the refresh rate, I am not aware of an API to actually **set** it 😛

**Comment by sa666666** [comment]
I'm assuming that if one selects a display mode with that resolution, then they will get that refresh rate.  I won't be doing anything more in Stella than allowing the user to select a different fullscreen resolution.  If it gives them the refresh they want, then great.  If not, too bad.  Either way, this issue will then be closed :smiling_imp:

**Comment by thrust26** [comment]
> While I know that SDL2 exposes the refresh rate, I am not aware of an API to actually **set** it 😛

SDL_GetClosestDisplayMode

**Comment by thrust26** [comment]
> While I know that SDL2 exposes the refresh rate, I am not aware of an API to actually **set** it 😛

SDL_SetWindowDisplayMode

**Comment by DirtyHairy** [comment]
Mmmmhhh, indeed 😏

**Comment by ale-79** [comment]
>[...]  I don't know of any LCD that does 50Hz, but I assume CRT will. But again, you would have to pick different modes until one happens to have 50Hz.

I think that at least in Europe, support for both 50Hz and 60Hz is a requirement for digital TVs (not monitors).

Here are the supported resolutions and framerates on an old laptop connected through HDMI to my Sony HD  TV:
![xrandr](https://user-images.githubusercontent.com/24704739/57883748-7005d980-7827-11e9-9583-5cb0f354f825.png)


I briefly tested setting the desktop resolution to 1920x1080 50Hz and playing some games (PAL and NTSC) in Stella. I didn't notice any issue with sound or game speed, and PAL50 games indeed look better in this way: they have smoother movement and show no beating effect in case of steady 25Hz flicker (e.g. Asteroids). 


Ideally, it would be nice to allow two different fullscreen modes for 50Hz or 60Hz games, and have Stella automatically choose which one to use based on the actual framerate of the game.

**Comment by thrust26** [comment]
@sa666666 
To get this going, I added logging of available full screen modes (verbose mode only) with a1d7315.

I briefly looked into the code and currently we are only using desktop modes. When I experimented with real full screen mode switching, that display got locked and I had to kill VS. Maybe you know better what to do here. 

Below is the list of supported modes of my monitor and TV. Anything between ~45Hz and ~65Hz might be useful to get closest to the real refresh rates of the current game. But we definitely have to filter these lists for the UI or use some automatic here.

Old Dell LCDs:
  0: 1280x1024@75Hz
  1: 1280x1024@60Hz (active)
  2: 1280x960@75Hz
  3: 1280x960@60Hz
  4: 1280x800@75Hz
  5: 1280x800@60Hz
  6: 1280x768@75Hz
  7: 1280x768@60Hz
  8: 1280x720@75Hz
  9: 1280x720@60Hz
  10: 1280x600@75Hz
  11: 1280x600@60Hz
  12: 1152x864@75Hz
  13: 1152x864@60Hz
  14: 1024x768@75Hz
  15: 1024x768@70Hz
  16: 1024x768@60Hz
  17: 800x600@75Hz
  18: 800x600@72Hz
  19: 800x600@60Hz
  20: 640x480@75Hz
  21: 640x480@72Hz
  22: 640x480@70Hz
  23: 640x480@60Hz
  24: 640x480@59Hz
  25: 640x400@75Hz
  26: 640x400@70Hz
  27: 640x400@60Hz
  28: 512x384@75Hz
  29: 512x384@60Hz
  30: 400x300@75Hz
  31: 400x300@60Hz
  32: 320x240@75Hz
  33: 320x240@60Hz
  34: 320x200@75Hz
  35: 320x200@60Hz

Samsung TV:
  0: 1920x1080@60Hz (active)
  1: 1920x1080@59Hz
  2: 1920x1080@50Hz
  3: 1920x1080@30Hz
  4: 1920x1080@29Hz
  5: 1920x1080@25Hz
  6: 1920x1080@24Hz
  7: 1920x1080@23Hz
  8: 1680x1050@60Hz
  9: 1680x1050@59Hz
  10: 1680x1050@50Hz
  11: 1680x1050@30Hz
  12: 1680x1050@29Hz
  13: 1680x1050@25Hz
  14: 1680x1050@24Hz
  15: 1680x1050@23Hz
  16: 1600x900@60Hz
  17: 1600x900@59Hz
  18: 1600x900@50Hz
  19: 1600x900@30Hz
  20: 1600x900@29Hz
  21: 1600x900@25Hz
  22: 1600x900@24Hz
  23: 1600x900@23Hz
  24: 1440x900@60Hz
  25: 1440x900@59Hz
  26: 1440x900@50Hz
  27: 1440x900@30Hz
  28: 1440x900@29Hz
  29: 1440x900@25Hz
  30: 1440x900@24Hz
  31: 1440x900@23Hz
  32: 1400x1050@60Hz
  33: 1400x1050@59Hz
  34: 1400x1050@50Hz
  35: 1400x1050@30Hz
  36: 1400x1050@29Hz
  37: 1400x1050@25Hz
  38: 1400x1050@24Hz
  39: 1400x1050@23Hz
  40: 1366x768@60Hz
  41: 1366x768@59Hz
  42: 1366x768@50Hz
  43: 1366x768@30Hz
  44: 1366x768@29Hz
  45: 1366x768@25Hz
  46: 1366x768@24Hz
  47: 1366x768@23Hz
  48: 1360x768@60Hz
  49: 1360x768@59Hz
  50: 1360x768@50Hz
  51: 1360x768@30Hz
  52: 1360x768@29Hz
  53: 1360x768@25Hz
  54: 1360x768@24Hz
  55: 1360x768@23Hz
  56: 1280x1024@75Hz
  57: 1280x1024@60Hz
  58: 1280x1024@59Hz
  59: 1280x1024@50Hz
  60: 1280x1024@30Hz
  61: 1280x1024@29Hz
  62: 1280x1024@25Hz
  63: 1280x1024@24Hz
  64: 1280x1024@23Hz
  65: 1280x960@60Hz
  66: 1280x960@59Hz
  67: 1280x960@50Hz
  68: 1280x960@30Hz
  69: 1280x960@29Hz
  70: 1280x960@25Hz
  71: 1280x960@24Hz
  72: 1280x960@23Hz
  73: 1280x800@60Hz
  74: 1280x800@59Hz
  75: 1280x800@50Hz
  76: 1280x800@30Hz
  77: 1280x800@29Hz
  78: 1280x800@25Hz
  79: 1280x800@24Hz
  80: 1280x800@23Hz
  81: 1280x768@60Hz
  82: 1280x768@59Hz
  83: 1280x768@50Hz
  84: 1280x768@30Hz
  85: 1280x768@29Hz
  86: 1280x768@25Hz
  87: 1280x768@24Hz
  88: 1280x768@23Hz
  89: 1280x720@60Hz
  90: 1280x720@59Hz
  91: 1280x720@50Hz
  92: 1280x720@30Hz
  93: 1280x720@29Hz
  94: 1280x720@25Hz
  95: 1280x720@24Hz
  96: 1280x720@23Hz
  97: 1280x600@60Hz
  98: 1280x600@59Hz
  99: 1280x600@50Hz
  100: 1280x600@30Hz
  101: 1280x600@29Hz
  102: 1280x600@25Hz
  103: 1280x600@24Hz
  104: 1280x600@23Hz
  105: 1152x864@75Hz
  106: 1152x864@60Hz
  107: 1152x864@59Hz
  108: 1152x864@50Hz
  109: 1152x864@30Hz
  110: 1152x864@29Hz
  111: 1152x864@25Hz
  112: 1152x864@24Hz
  113: 1152x864@23Hz
  114: 1024x768@75Hz
  115: 1024x768@70Hz
  116: 1024x768@60Hz
  117: 1024x768@59Hz
  118: 1024x768@50Hz
  119: 1024x768@30Hz
  120: 1024x768@29Hz
  121: 1024x768@25Hz
  122: 1024x768@24Hz
  123: 1024x768@23Hz
  124: 800x600@75Hz
  125: 800x600@72Hz
  126: 800x600@60Hz
  127: 800x600@59Hz
  128: 800x600@50Hz
  129: 800x600@30Hz
  130: 800x600@29Hz
  131: 800x600@25Hz
  132: 800x600@24Hz
  133: 800x600@23Hz
  134: 720x576@60Hz
  135: 720x576@59Hz
  136: 720x576@50Hz
  137: 720x576@30Hz
  138: 720x576@29Hz
  139: 720x576@25Hz
  140: 720x576@24Hz
  141: 720x576@23Hz
  142: 720x480@60Hz
  143: 720x480@59Hz
  144: 720x480@50Hz
  145: 720x480@30Hz
  146: 720x480@29Hz
  147: 720x480@25Hz
  148: 720x480@24Hz
  149: 720x480@23Hz
  150: 640x480@75Hz
  151: 640x480@73Hz
  152: 640x480@60Hz
  153: 640x480@59Hz
  154: 640x480@50Hz
  155: 640x480@30Hz
  156: 640x480@29Hz
  157: 640x480@25Hz
  158: 640x480@24Hz
  159: 640x480@23Hz
  160: 640x400@60Hz
  161: 640x400@59Hz
  162: 640x400@50Hz
  163: 640x400@30Hz
  164: 640x400@29Hz
  165: 640x400@25Hz
  166: 640x400@24Hz
  167: 640x400@23Hz
  168: 512x384@60Hz
  169: 512x384@59Hz
  170: 512x384@50Hz
  171: 512x384@30Hz
  172: 512x384@29Hz
  173: 512x384@25Hz
  174: 512x384@24Hz
  175: 512x384@23Hz
  176: 400x300@60Hz
  177: 400x300@59Hz
  178: 400x300@50Hz
  179: 400x300@30Hz
  180: 400x300@29Hz
  181: 400x300@25Hz
  182: 400x300@24Hz
  183: 400x300@23Hz
  184: 320x240@60Hz
  185: 320x240@59Hz
  186: 320x240@50Hz
  187: 320x240@30Hz
  188: 320x240@29Hz
  189: 320x240@25Hz
  190: 320x240@24Hz
  191: 320x240@23Hz
  192: 320x200@60Hz
  193: 320x200@59Hz
  194: 320x200@50Hz
  195: 320x200@30Hz
  196: 320x200@29Hz
  197: 320x200@25Hz
  198: 320x200@24Hz
  199: 320x200@23Hz

**Comment by thrust26** [comment]
> Just going over this again. Should I add a dropdown box that allows to select the fullscreen video mode? 

Some time ago I prepared such a drop down box in the VideoDialog already. I wonder if it makes sense to let the user decide or if Stella should do that instead. And the user only has the option to decide what he wants Stella to do (switch to full screen mode, chose the best matching mode).

**Comment by sa666666** [comment]
If we're going into that level of detail (allowing user to decide vs. Stella, only showing certain modes, validating modes, etc), then this should be moved to 6.2.

**Comment by thrust26** [comment]
For sure. Or later, depending on the effort needed.

**Comment by gulp437** [comment]
Moved to year 2034 for Stella 8.9  aye aye sir!

**Comment by gulp437** [comment]
Very good to not fool PAL-users any longer and finally tell them the truth now, that this thing here was never planned and will never come. Correctly, at last. Very good to go this step now, "stephena".

I mean, it's not a NTSC-related problem, right. Therefore i see no need to hurry up anyway. Better talking about it the next three years and then, before people will find out the truth, postpone it again and again and again. I am sure this will work, we just should make sure, that the PAL-users don't find out.

I mean, people which demanded this "Hz must fit FPS" things are all chasing butterflies. Same as audiophiles, that swear on golden fuses. *lol* Those people can hear the grass grow, you know, hahaha. Let's go on with our old STELLA-rule: "NTSC first NTSC first NTST first". Good that the true colors now comes out and we speak our mind clearly now, that this will NOT COME. Finally.

Now all PAL-users know the truth about this point here and that it is on lowest priority, like all other PAL-related things in STELLA. I mean, look at the PAL color-palette and how we fooled the PAL'ers for years with unrealistic colors *lol*. Can you remember "stephena", how funny it was, not to include a correct one on purpose for all those years *lol*. We often made jokes about it, right?

And there are other things too, which prevented this point here and i want to speak openly about them now. Listen carefully, when you are a PAL-user. There is alot of snow in the area in which i live here, which is another point, why it is absolutely NOT possible to work on PAL-related things in STELLA now and in next years. Only on NTSC-related things in STELLA can be worked, when there is alot of snow. Everybody knows this, it's a very old rule from the STELLA book of rules, which we get from our developer-predecessors and which is from the middle-age. Of course we stick to this old book. Wow, now when this snow thing was spoken out, i feel much better.

And the last point is this. Also this claim, that there exists some 50 FPS compatible flatscreen-monitors or TV's in Europe are nothing more than bad FAKE news. Nobody there has a monitor or TV which can do someting else than 60Hz, believe me. One reason more, to never include this.

Thanks.

**Comment by sa666666** [comment]
We still plan to implement this feature as time allows.  But I am through with responding to abusive people.

**Comment by gulp437** [comment]
Well, "stephena", why not being honest finally. This is your chance now, even when it seems to be hard for you. Why not telling PAL-users that these things are on lowest priority they could be?

**Comment by thrust26** [comment]
I checked the old code from 3.9.3 and there is **no** logic regarding NTSC, PAL or framerates. It just queries SDL for the available fullscreen modes and then selects the first(!) one which fits to the requirements. In case of 'fullres=AUTO', this would be the display size at minimal zoom (640x480 for NTSC, a bit larger for PAL due to scaling). Else the resolution requested by `fullres=WxH` (if larger than the minimal requirements). The former explains why by default for NTSC 640x480 was used and 800x600 for PAL. The different framerates resulted from OS settings. So this would not help us today at all. 

Now, here is my idea for today:
- The user can only request the **minimal** resolution used for fullscreen. So that he can decide between maximum speed and maximum quality. The list of resolutions would be created by Stella, listing all available resolutions starting e.g. at 640x480.
- Stella will then automatically select the mode which comes closest to 60 or 50Hz (or multiples?) and has a minimal sufficient resolution.

Remaining problems:
- OS reaction to changing resolution, especially when Stella should not exit properly and we get stuck in the fullscreen mode. (Note: this can be **very** annoying during development, especially when you have only one monitor)
- In case of multiple monitors, which monitor defines the resolution list? Both combined? Their intersection? The current one used by Stella? We definitely need a fallback if the current monitor does not support the minimal resolution requested by the user.
- What happens with the UI? Debugger, launcher, dialogs? Currently they are also displayed in fullscreen mode, which makes sense, because they are always using maximum monitor resolution. And the code makes sure that everything fits. 
But what if the user selects e.g. 640x480 for fullscreen? The UI breaks. E.g. the debugger wouldn't work anymore. Also messages and dialogs using a larger font would not work in low resolution fullscreen modes. Should we switch back and forth between real fullscreen and the current one? And/or use different fonts for real fullscreen modes? Or should we limit the minimal fullscreen resolution based on the current fonts? Or...?
- Anything I forgot?

**Comment by sa666666** [comment]
Yep, there was nothing in the old code that did what we want.  IMO, it was just a fluke that it worked at all.  IOW, we were relying on behaviour of SDL that is no longer happening.

There's a reason this issue has been open for so long.  You've listed quite a few of the reasons.  This is not as easy to fix as just adding a 'button in the UI', as someone suggested.  And TBH, seeing all the potential bugs that we'd have to deal with, I now wonder whether we should implement it at all.  In the end, is the functionality we get going to be worth the effort??

Regarding 'mode changes', I specifically went with `SDL_WINDOW_FULLSCREEN_DESKTOP` when opening the display, since in SDL1 there were all kinds of bugs when opening in a smaller resolution:
- If the app crashes, the desktop is not restored
- In some cases, even if the app exits normally, the desktop icons are all 'squished' into the upper-left corner.
These were bugs reported in the past, and going to SDL2 and `SDL_WINDOW_FULLSCREEN_DESKTOP` is what fixed it.  I don't want to regress.

In any event, if you will be working on this, please do it in a branch first.  I don't want to break master with this potentially very invasive change.

**Comment by thrust26** [comment]
Fully agreed on all points. 

Actually I have already created a branch for experimenting with it. :smile:

**Comment by sa666666** [comment]
Many (most?) games/emulators/etc in this day and age use the desktop resolution for everything.  Even if the computer isn't fast enough to render at that res, since the video hardware on all systems is now accelerated, we can use scaling.  For example, if a computer isn't fast enough to render at 4K resolution, that doesn't mean we can't use that resolution.  What happens is that we open the fullscreen window at 3840x2160, but then only render into a small buffer of, say, 1920x1080.  And then just let the hardware scale to the higher resolution.

This is actually what Stella does now.  SDL has built-in functions to do it, but I abstracted that away and have a concept of a `FBSurface` that can be scaled as needed.  And the SDL developers themselves recommend doing things this way too.  We are now (mostly) in a world of hardware-accelerated video and LCD-based computers.  Those have a set resolution, and the best alternative is to use that resolution at all times _for opening the fullscreen window_.  Note that doesn't mean we have to render at that res, only that we have to open at that res.

So IMO, the performance issue is a red herring.  We shouldn't worry about that part.  Really, what this issue is addressing is wanting 50Hz fullscreen mode for PAL games, and 60Hz for NTSC ones.  It would be easy enough to just do that part, I think.  The remaining problems just go away when we assume at standard (desktop) resolution for fullscreen mode.

I would suggest concentrating on only the 50 vs. 60 Hz issue.  And perhaps try out some code for that first.  Because if we can't get that working correctly, then the rest is a waste of time anyway.

**Comment by sa666666** [comment]
I just realized that my last post basically invalidates this issue completely :smiling_imp: I no longer feel that we need to offer the user the ability to change fullscreen mode, because of what I said above.  Now, offering the ability to use 50Hz or 60Hz variants of that fullscreen mode may be an option.  But I wonder if that should be a separate issue??

**Comment by thrust26** [comment]
Funny, I had the same idea. Just switching between 50 and 60Hz. I wonder how common 50Hz support for the native resolution is. What do(es) your monitor(s) say? (System log in verbose mode).

**Comment by sa666666** [comment]
```
Supported video modes for display 0:
  0: 3840x2160@60Hz (active)
  1: 3840x2160@50Hz
  2: 3840x2160@30Hz
  3: 3840x2160@25Hz
  4: 3840x2160@24Hz
  5: 2560x1440@60Hz
  6: 1920x1200@60Hz
  7: 1920x1080@60Hz
  8: 1920x1080@50Hz
  9: 1920x1080@30Hz
  10: 1920x1080@25Hz
  11: 1920x1080@24Hz
  12: 1680x1050@60Hz
  13: 1600x1200@60Hz
  14: 1600x900@60Hz
  15: 1440x900@60Hz
  16: 1280x1024@75Hz
  17: 1280x1024@60Hz
  18: 1280x800@60Hz
  19: 1280x720@60Hz
  20: 1280x720@50Hz
  21: 1152x864@75Hz
  22: 1024x768@75Hz
  23: 1024x768@60Hz
  24: 832x624@75Hz
  25: 800x600@75Hz
  26: 800x600@60Hz
  27: 720x576@50Hz
  28: 720x480@60Hz
  29: 720x400@70Hz
  30: 640x480@75Hz
  31: 640x480@60Hz
```
I believe most monitors will have a 50Hz variant of the native resolution, but if not then it's easy to detect that (and in which case, we do nothing).

**Comment by thrust26** [comment]
We could simply add an option which says something like "try to adapt refresh rate". If it works it works, if not it doesn't. And maybe show a message.

**Comment by sa666666** [comment]
Yes, that's what I'm thinking.  If it "doesn't work", then it does whatever Stella is doing right now.  Otherwise it switches to a different version of the desktop resolution.  So what we need is:
- Figure out what 'best refresh rate' means when the game is neither exactly 60Hz nor 50Hz.  Perhaps only games displaying at 50Hz or less should use that mode, else it uses the 60Hz variant?
- Determine whether it actually makes a difference.  For now, we're assuming that opening a 50Hz game in a 50Hz mode will result in less jitter/vsync issues/etc.  But we don't know until we try it.

**Comment by thrust26** [comment]
Many if not most 2600 games do not create exactly 60 or 50Hz. Is a refresh rate as close as possible really always better?

E.g. I wonder if 75Hz like my old monitors have would be better for PAL than 60Hz. A factor of 1.5 sounds better than 1.2, no?

**Comment by sa666666** [comment]
That's exactly my point; we don't know what will work best.  I've only used 60Hz screens for the past 10 years, and before that it would have been CRTs with 75Hz or more.  I don't recall how 75Hz worked with Stella, and I have no experience whatsoever with < 60Hz, we need to experiment before we can finalize any code.

Personally, I don't think we should optimize our use-cases for CRT.  That is a dead technology that's not coming back.  Going forward in fact, the trend is for refresh rates to increase (120/144/240Hz is becoming more common).

**Comment by thrust26** [comment]
I wasn't referring to CRTs. I meant that e.g. 61 Hz might look worse at 60Hz than at a higher refresh rate. Every ~1s one frame has to be skipped then. If we display at e.g. 75Hz, we would double every 4th frame instead of skipping. So what would look better? 

Maybe simply using the highest possible framerate is better than trying to match an slightly off framerate as close as possible.

**Comment by sa666666** [comment]
Yes, it's possible that there's no improvement at all, or that it actually gets worse.  What happens on a real console and CRT when the framerate deviates from the normal 50/60Hz?  I know if the scanlines go too out of whack that the screen rolls.  But before that, do we get 'judder' or irregular movement??

**Comment by thrust26** [comment]
Not with the 2600, the games will just run slower or faster. 

On US TV shows played at PAL especially pan shots create a lot of very noticeable judder. That's what would happen when a 60Hz game is played at 50Hz.

**Comment by sa666666** [comment]
At this point I guess there's nothing else to say, other than to try it.  It doesn't make sense to implement the feature if it doesn't improve anything, so we need to see visually if it will make a difference ...

**Comment by thrust26** [comment]
I will try to create something simple and then we can test.

**Comment by DirtyHairy** [comment]
Late to the party 😏 I don't have anything to add to the discussion that hasn't been said already anyway. I think switching between 50Hz and 60Hz automatically might make kinda sense, but my feeling is that "real" screen modes that correspond to a hardware state and not are just a virtual surface are quickly becoming a thing of the past, and this whole topic can easily become a platform dependent rabbit hole.

**Comment by thrust26** [comment]
True. But unlike modern games, the old games were programmed to look best on old hardware. So emulating old hardware bit more closely makes sense.

**Comment by sa666666** [comment]
I think the only way to match old hardware is if the display is v'synced to the framerate, whatever that may be.  And since LCDs don't work like CRTs in that fashion, I don't see this ever really being possible.  We can perhaps make it look a little better, but we'll never be able to adequately emulate a variable-rate screen on a fixed-rate monitor, and have it as smooth as the real thing.

I've followed other emulators a little, and I've noticed this problem is never really solved.  I think this is because it fundamentally _can't_ be solved.

**Comment by thrust26** [comment]
Oh, it can be solved if we sync emulation to refresh rate. But I doubt that anyone would like that. 😈

**Comment by sa666666** [comment]
No, I mean on modern LCDs there are only one or two native refresh rate (60Hz or 50Hz).  Anything else requires a mode change.  I don't see how to run a 50Hz game on a 60Hz screen and not have judder.

TV's/Bluray players/etc have been struggling with this for years.  The solution they came up with is to have a refresh rate that is a common multiple of both, or use various tricks like skipping a frame every so often.  And it still isn't a solved problem.  A lot of TV's have anti-judder processing being done, since it's still a problem displaying content on a device with a different refresh rate.

**Comment by DirtyHairy** [comment]
> Oh, it can be solved if we sync emulation to refresh rate. But I doubt that anyone would like that. 😈
> 
> Though, maybe it is acceptable for very low differences?

That would require considerable changes to the way audio and emulation speed are currently handled.

**Comment by thrust26** [comment]
I was afraid you would say that. :smile:

**Comment by DirtyHairy** [comment]
Au contraire, it saves you a lot of frustrating work 😛

**Comment by thrust26** [comment]
No big deal, I am bored anyway. :smile:

**Comment by thrust26** [comment]
Updated plan: Add a simple checkbox (e.g. "Align display refresh rate). If it is checked, Stella will switch to the video mode which has the current display resolution and the refresh rate closest to the game's TV type. 

Questions: 
- We could do that for non-fullscreen modes too. Agreed?
- When do we switch back to the original refresh rate? Exiting to the launcher? Exiting Stella? Switching to the debugger?

**Comment by thrust26** [comment]
Hm... I seems that we can only change the display settings (incl. refresh rate) in **real** fullscreen mode (`SDL_WINDOW_FULLSCREEN`).  Currently Stella always uses a desktop fullscreen mode (`SDL_WINDOW_FULLSCREEN_DESKTOP`). 

Since we only want to change the refresh rate it seems less critical to use the real fullscreen mode. And we can limit using it to cases where we really change the refresh rate. 

What do you think?

**Comment by sa666666** [comment]
Yep, this is what I was referring to when I said always using the desktop res.  I didn't just mean I selected it, but specifically that I instruct SDL by using a different flag (aka FULLSCREEN_DESKTOP).  As mentioned, this is what is recommended by SDL developers for all future development.  It (among other things) guarantees that a video mode change doesn't happen, you don't get rearrangement of icons, etc.

You are right that if you want to mess with 'real' fullscreen modes, you have to change the flag to `SDL_WINDOW_FULLSCREEN`.  I don't mind doing that, but only when the user selects to do so, by selecting the option you're adding.  By default, I don't want it to happen.

Actually, several people over the past year have asked about variable-rate refresh monitors, and to support that we would need to go the route you're now thinking about.  So my opinion is this:
- I don't mind having this functionality, as long as it defaults to off, and doesn't break any functionality that we now have.
- This only applies to fullscreen modes.  Windowed modes can't have a separate refresh rate.

**Comment by thrust26** [comment]
You can check my latest commit to feature-fullscreen. Enabling the real FS mode is currently not very elegant and causes unnecessary extra flicker, but it works.

**Comment by sa666666** [comment]
Yep, mode changing will always result in this flickering, as the monitor physically takes some time to change to a new mode.  This approach was quite common with SDL1, but is mostly gone with SDL2.  I don't think there's any way around this, unfortunately.

I don't know how to name it.  As we've seen, my choices for names is not always very good :smiling_imp:

**Comment by thrust26** [comment]
I went with "adapt". :smile:

Can you please test with other OS? Especially MacOS, because it has its own code.

**Comment by sa666666** [comment]
OK, here's the results of my testing so far, along with comments on what I think is going wrong in some cases.  For the rest of the message, I will refer to the new feature as _adapt_.  All testing was done with Swoops, as it's a horizontal scroller which clearly shows judder in PAL mode.

**Windows**

- Turning adapt on makes Stella not have any sound!  I think this is an interaction between a screenmode change happening, and the sound system not being re-initialized.  This doesn't happen on Linux or Mac.
- I can't seem to get a 50Hz mode to come up.  I verify this by the popup message that appears when toggling fullscreen/windowed mode.  It always says 60Hz.  But there is a pause in the mode switching, so it does seem to be doing _something_.  I think it's because of the modes my monitor has, and it doesn't have the desktop resolution mode in 50Hz, only 60Hz.  There _are_ 50Hz modes, but they're all smaller than the desktop mode.  Perhaps the code needs to be extended to select one of them??

**Mac**

- On my system, only 60Hz modes are available.  This could be due to it being a Hackintosh, or perhaps Macs don't present 50Hz modes at all.  But as with Windows, there is a delay in switching modes, so something is happening.  I think @DirtyHairy will need to test on a real Mac.

**Linux**

- This is the only place where (a) the desktop res has both 50 and 60Hz modes, and (b) it properly switches between them.  I can tell this from both the delay in switching, and the message that pops up onscreen.  However, I don't see any difference between adapt on or off.  That is, the judder is still there.

So that's all I have for now.  Mac is inconclusive, Windows still needs more work, and in Linux it doesn't seem to make any difference.

**Comment by thrust26** [comment]
Thanks for testing. Funny, I have tested with SWOOPS! too. :smile: 
But only the switching. I cannot test the judder with my monitors, because they do not support 50Hz (I switched NTSC to 75Hz instead) I will try with other systems tomorrow.

> Turning adapt on makes Stella not have any sound! I think this is an interaction between a screenmode change happening, and the sound system not being re-initialized. This doesn't happen on Linux or Mac.

I cannot reproduce the problem on my system. So it will be hard to fix for me.


> Sometimes I get a 60Hz mode, and sometimes 59Hz?

Your monitor doesn't support 59Hz, does it? Could be an SDL problem or maybe a rounding error. The code only asks for 60 Hz.

> I can't seem to get a 50Hz mode to come up. 

If your monitor doesn't support 50Hz in the current desktop resolution, then the code will not switch. Changing the display size seems too risky to me. What do you think about it?

> On my system, only 60Hz modes are available. This could be due to it being a Hackintosh, or perhaps Macs don't present 50Hz modes at all. But as with Windows, there is a delay in switching modes, so something is happening. I think @DirtyHairy will need to test on a real Mac.

Just ask your monitor. It should display the current refresh rate, no?

> That is, the judder is still there.

That would be bad. But we have to make sure that the monitor is at the correct setting. So it would really help is you double check what the monitor tells you.

**Comment by thrust26** [comment]
I just tested SWOOPS! /Splatfrom PAL with my TV. And it definitely looks **much** smoother at 50Hz.

Since the TV has a different resolution than my monitors, I noticed some positioning and sizing problems. But these seem unrelated to the new fullscreen mode and happen with the desktop fullscreen mode too. Definitely something to look into in a different issue.

**Comment by thrust26** [comment]
BTW: The amount of flicker could be reduced (or eliminated) if we wouldn't destroy the renderer each time. Wouldn't it be sufficient to destroy the renderer only if `myWindow` is destroyed (or the renderer type has been changed)?

The comment in the code is very vague: `// Always recreate renderer (some systems need this)`

**Comment by thrust26** [comment]
I committed a change, where the renderer is only recreated when necessary. This reduces (or even eliminates) flicker in real fullscreen mode.

**Comment by sa666666** [comment]
For Linux, I'm able to confirm that switching to a 50Hz mode is actually happening on the monitor (checked the info from the monitor settings, and it says displaying in a non-optimal 50Hz mode).  But I still don't see much (any?) difference in the juddering in Swoops.  I'm content to say it's working if you're seeing a difference, but for now, I can't.  I want to get it working in Windows, so I can see if it makes a difference there.

I'm going in to my office sometime this week to borrow a real Mac, and then I can do some testing on real hardware for that platform.  BTW, looks like we won't be going to back to on-campus classes this Fall, so I'm working from home until at least December.

Regarding the comment above, I believe Macs required to re-create the render each time.  But that was under older versions and OpenGL.  The Mac port is now using a newer SDL version and Metal (instead of OpenGL).  So this may no longer be a requirement.  Again, all the more reason why I need to test on a real Mac.

**Comment by sa666666** [comment]
The "don't re-create renderer" code works fine in Linux and Windows for me.  Still need to test on a real Mac.

**Comment by thrust26** [comment]
> But I still don't see much (any?) difference in the juddering in Swoops!

Did you try Splatform? It is pretty obvious there.

**Comment by sa666666** [comment]
Yes, that's what I was using.  I don't suppose you can find a way to record (in 50Hz) what you're seeing, to confirm whether what I'm seeing is correct?

**Comment by thrust26** [comment]
I don't think I can, no.

Did you compare 50Hz vs. 60Hz? PAL at 50Hz is as smooth as the NTSC version at 60Hz, But with PAL at 60Hz you notice the judder. 

Maybe your display is playing a role here?

**Comment by sa666666** [comment]
OK on the video.  I don't know how to do it either (other than pointing a camera at the monitor and recording that way).

I would expect to see what you're explaining (PAL at 50Hz on a 50Hz screen is just as smooth as NTSC at 60Hz on a 60Hz screen).  But I'm definitely not seeing that.

Do you have vsync on or off in this case?  I always have it on.  Note that on my Windows system I still can't get a 50Hz mode.  I think I need to hook up another monitor to try it.  It's at work now; I'm going out there in the next few days.

For now, the only system where I can get a 50Hz screen is Linux.  And it doesn't seem to make a difference there.  But there are sometimes issues with video and Linux, so that doesn't mean the code doesn't work.  I really need to test in Windows, where I suspect it will work fine.

**Comment by thrust26** [comment]
I hope @DirtyHairy can help us here a bit.

**Comment by DirtyHairy** [comment]
Nothing useful on my Macbook, sorry.

If I don't have an external display connected, then Stella briefly exits and reenters fullscreen mode as it tries to switch modes. However, the refresh rate seems unchanged. Actually, I don't think the Mac ever switches mode on the internal panel, it only creates virtual surfaces of the desired resolution that are scaled to the physical size.

If I have my external 4k display connected, then can't get Stella to display on the external display from the start. If I drag Stella over to the other display and enable "adjust", the switching "modes" will end up with a fullscreen window that is scaled beyond the screen margins on the internal display. In a mirrored setup, the effect is the same is if I am running only in the internal panel --- no change in refresh rate.

Actually, I don't think I have ever seen an application that really switches video modes on a Mac. Games change "resolution", but I think that only creates a virtual surface that is upscaled. You can configure video modes when you setup the screen (and even that not for all displays --- I can do so on my TV and on VGA displays, but not on my 4k), but I doubt that this is exposed to applications.

**Comment by DirtyHairy** [comment]
For the notes: I can also see the stutter when running PAL on a 60Hz display.

**Comment by sa666666** [comment]
> Actually, I don't think I have ever seen an application that really switches video modes on a Mac. Games change "resolution", but I think that only creates a virtual surface that is upscaled. You can configure video modes when you setup the screen (and even that not for all displays --- I can do so on my TV and on VGA displays, but not on my 4k), but I doubt that this is exposed to applications.

Yep, that's what Stella has been doing all along, and what is recommended by SDL.

I think I know the issue in Linux now.  The AMD drivers are setting v'sync/60Hz at the window manager level.  So it doesn't matter if Stella requests (and actually gets) a 50Hz mode.  It is still vsync'ed at the 60Hz rate.  I really don't want to screw around with this.  I have smoothly updating video in Linux after some amount of work, and don't want to mess it up with experimentation.

I will test on an iMac by the end of the week, once I get the system from work.  But I suspect it will be the same as @DirtyHairy results.  Maybe this will only end up working in Windows, and we would expose it on that system only??

**Comment by thrust26** [comment]
If the AMD drivers are the problem, then we might have problems with Windows too, no? Would disabling VSYNC help here?

**Comment by thrust26** [comment]
@DirtyHairy 
Have you tried switching your refresh rate manually?
https://support.apple.com/de-de/HT210742

Do that improve quality for PAL games?

**Comment by sa666666** [comment]
No, it's not a problem with the AMD drivers, but with the Linux graphics subsystem in general.  Basically I have an option turned on called 'TearFree' that makes sure there's no tearing in the various graphical updates.  This is done at the OS level.  So even if you select a different refresh rate or turn off vsync, it still enforces it.  So I can't test on this machine.  And I don't want to screw around with it, since it benefits all the other apps I use, and it took so long to get it all working.

I can test on another Linux machine, but honestly, I see this feature maybe only ending up working on Windows.  It's also tangentially related to variable rate refresh (VRR), and that's something that mostly only works on Windows too.  So again, this may end up being a Windows-only feature.

**Comment by thrust26** [comment]
But does every Linux user have this option turned on?  There are FreeSync these monitors now, which are used by gamers, and they can sync to a wide range of refresh rates. These must work with Linux too, no?

**Comment by sa666666** [comment]
I suppose not _every_ Linux user has that option turned on, but if you look online at many HOWTO's, they all recommend to do it.  And VRR monitors are only supported by certain video drivers under Linux.

As @DirtyHairy has said in the past, the video card/driver situation in Linux is not ideal, and not at all the same as in Windows.  Realistically, while we can make sure the code doesn't break anything in Linux, I don't really see it being as widely used as it would be for Windows users.  Beyond the other LInux system I have to get a monitor for, I don't have any other way to test this in Linux.  We will need to release betas if we expect some feedback.

But before that even happens, I would like to make sure it works in Windows.  Windows has the best support for video cards and the widest usage from Stella users, so if we can't get it working reliably there, I don't really see the point of the feature.

**Comment by thrust26** [comment]
We can concentrate on Windows, no problem for me. :wink: There it seems to work fine by me, but my tests are limited. I will ask around in the club, maybe we find some more testers there.  Fingers crossed.

And we probably should add some additional documentation, which explains the limitations. Then we can use "RTFM!" whenever someone comes around asking why it isn't working for him. 😈

**Comment by DirtyHairy** [comment]
> @DirtyHairy
> Have you tried switching your refresh rate manually?
> https://support.apple.com/de-de/HT210742
> 
> Do that improve quality for PAL games?

Doesn't work, I cannot change refresh rate on my laptop panel or on  my 4k display.

**Comment by sa666666** [comment]
I'm willing to bet that if you turn on verbose logging and look at the available video modes, then none of them will be 50Hz.  That's what it's like on my Hackintosh, but because this isn't a genuine Mac I can't use it for testing this stuff.  But I'm willing to bet a _real_ Mac doesn't have 50Hz modes at all.

**Comment by thrust26** [comment]
Nanny-OS! 😛

**Comment by thrust26** [comment]
https://atariage.com/forums/topic/307057-adapting-display-refresh-rate/?do=findComment&comment=4542644

**Comment by sa666666** [comment]
OK, found some time to check on a real iMac.  Like @DirtyHairy said, this machine doesn't expose 50Hz modes at all.  So not sure we can do anything here.

I still need to set up a 1080p monitor on my Windows development machine.  It looks like the res I run at (2560x1440) doesn't have a 50Hz equivalent, and since the code only switches modes when there's  an exact res available, I can't test with this monitor.

I think it also has to do with VESA modes.  Some of them have 50Hz varieties and some don't.  1080p definitely does, so it should be fine there.  Some other ones may not have a 50Hz mode for anyone, so they may never be able to take advantage of the feature.

**Comment by sa666666** [comment]
Maybe we need to show a message in the UI telling if a 50Hz mode is available or not.  It might help reduce complaints/feedback if the user sees that 50Hz simply isn't available for their screenmode.

**Comment by thrust26** [comment]
Either that, or we allow the code to change resolution.

**Comment by thrust26** [comment]
@sa666666 Should we take the risk of changing resolution for fullscreen? Then e.g. your display would switch.

**Comment by sa666666** [comment]
It would have to change only for TIA mode (not for debugger, ROM launcher, etc).  And it would have to be to one that is closest but a little smaller.  But I suppose we could at least try it.

Since R77 is 1280x720, and that mode typically does have 60Hz and 50Hz available, yes, I believe it could be possible.  One unknown is how a typical TV will deal with 50Hz, even if it reports supporting it??

**Comment by sa666666** [comment]
There's also something else we need to work on, but it's not high priority right now.  On at least one of my systems, the default for the monitor is 59Hz, not 60Hz (even though both are available).  So that means in the code, we really shouldn't hardcode '60', but instead use whatever the default is for that monitor.  So we'd need to query that value, and then use that value instead of '60'.

**Comment by thrust26** [comment]
> It would have to change only for TIA mode (not for debugger, ROM launcher, etc)

I wouldn't do that as this will result into constant switching back and forth between modes. The current code switches in emulation mode only and then keeps the mode until a ROM requires a different refresh rate.

> And it would have to be to one that is closest but a little smaller. But I suppose we could at least try it.

Yes, I would search for the closest resolution. Not sure if we should limit only to smaller ones.

> There's also something else we need to work on, but it's not high priority right now. On at least one of my systems, the default for the monitor is 59Hz, not 60Hz (even though both are available). So that means in the code, we really shouldn't hardcode '60', but instead use whatever the default is for that monitor. So we'd need to query that value, and then use that value instead of '60'.

I don't think this will work. The game determines the requested frequency, not the display. So if your display supports 59 and 60 Hz modes, an NTSC game will always use the latter. Which is exactly what the change should do.

**Comment by sa666666** [comment]
> > It would have to change only for TIA mode (not for debugger, ROM launcher, etc)
> 
> I wouldn't do that as this will result into constant switching back and forth between modes. The current code switches in emulation mode only and then keeps the mode until a ROM requires a different refresh rate.

OK, I just noticed that in testing.  Locking to the last-used frequency probably results in less changes overall.  So that's fine, I guess.

> > And it would have to be to one that is closest but a little smaller. But I suppose we could at least try it.
> 
> Yes, I would search for the closest resolution. Not sure if we should limit only to smaller ones.

Well, the desktop is typically the largest resolution being used, so there won't typically _be_ a larger mode.  I suppose we could look in both directions, and figure out which one is closer.

> > There's also something else we need to work on, but it's not high priority right now. On at least one of my systems, the default for the monitor is 59Hz, not 60Hz (even though both are available). So that means in the code, we really shouldn't hardcode '60', but instead use whatever the default is for that monitor. So we'd need to query that value, and then use that value instead of '60'.
> 
> I don't think this will work. The game determines the requested frequency, not the display. So if your display supports 59 and 60 Hz modes, an NTSC game will always use the latter. Which is exactly what the change should do.

I'm trying to keep the functionality between non-adaptable (current code) and adapable (new code) the same as much as possible.  Here's the use case:
- Right now, the default (native) screenmode for my Windows machine is 2560x1440, 59Hz.  So NTSC and PAL ROMs open in that mode in fullscreen, when not adapting to refresh rate.  There is no mode change.
- Your code returns '60' for NTSC and '50' for PAL, hard-coded.
- The problem is, the new code will ask for 60Hz for NTSC and 50Hz for PAL.  So it will do a mode change for both types, since 59 is not equal to either.

Basically, some monitors have an NTSC compatibility setting that exposes 59Hz _and_ 60Hz.  I would argue that when switching to 60Hz, we really should be switching to the mode that Stella/Windows is actually using on the desktop.  With the current code, I'm getting mode switches from windowed to fullscreen no matter what the ROM is asking for, since both 60Hz and 50Hz are different from 59Hz.  So the worst of both worlds.  It should be treating the 60Hz as 59Hz.  Or more to the point, using whatever the Windows desktop is currently using.

**Comment by thrust26** [comment]
> I would argue that when switching to 60Hz, we really should be switching to the mode that Stella/Windows is actually using on the desktop. 

Why? The game runs at 60Hz and we want to avoid judder. So switching to 59Hz (or keeping it at that) makes no sense. And yes, the mode should switch in both cases if there are better refresh rates available. That's its purpose.

**Comment by sa666666** [comment]
Then it's a regression, though.  Before this code was added, toggling fullscreen mode didn't do a mode switch; now it does.  And the 59/60Hz modes are actually the same behind the scenes, but SDL doesn't know that, and switches from one to the other even though they are eventually the same. I've read up on this a little over the past hour, and it seems like what's sent to the video card is the same no matter what.  In fact, it's the 59Hz mode that is more correctly (actually 59.94), not the 60Hz one :smiling_imp: 

I guess we can document that the user has to update their Windows settings to make sure 60Hz is the default, not 59Hz.  But requiring this is again, by definition, a regression.

**Comment by thrust26** [comment]
No, it only switched if the checkbox is checked. Else no refresh rate should ever happen.

**Comment by thrust26** [comment]
BTW: I wonder if we should select 59 (aka 59.94 Hz) instead of 60 Hz. IIRC the 2600 creates a frame rate slightly lower than 60 Hz (1,193,182 / 76 / 262 = 59.92). And maybe 49 Hz for PAL.

Unfortunately SDL only returns integers for the refresh rates.

**Comment by sa666666** [comment]
Yes, that's the number I'm referring to above.  Apparently, what is sent to the video card is 59.94Hz regardless of whether 59Hz or 60Hz is selected.  The former is the true rate (for TV's), and the latter is for compatibility (with computer monitors).

IMO, we should be looking at the list of modes (for a particular res) and using whatever is closest to what is already in use.  That's why I think the code should ask for the 59Hz rate on my system, even if 60Hz is available.  From the POV of refresh rate sent to the monitor, both eventually do exactly the same thing.  But from the POV of SDL, one will cause a mode change and the other won't.

I guess my main point is to look at the numbers dynamically, instead of hardcoding 60 and 50 statically.

**Comment by thrust26** [comment]
Are you 100% sure that 60 Hz are **always** actually 59.94 Hz?

So when SDL reports the current mode is 59 Hz, then this means 59.94 Hz. And when it reports 60 Hz, then it means the same? 

We could add a tolerance of -1 Hz to the code (59/60 and 49/50 Hz). But only if we are sure what we are doing.

**Comment by thrust26** [comment]
https://support.microsoft.com/en-us/help/2006076/screen-refresh-rate-in-windows-does-not-apply-the-user-selected-settin

Looks like you are correct. But there seems to be nothing adequate for PAL. Weird.

**Comment by sa666666** [comment]
Yep, that's pretty much the same page I was looking at.  So the refresh rates are the same, and if one is within 1Hz of the other, then we can consider them equal.  But since they are both equal and we have to choose one of them, I advocate to change to the one that the desktop is running at.  So on my system using 59Hz on the desktop, we switch to 59Hz fullscreen, not 60Hz.  This eliminates the mode change and at the same time uses a refresh rate exactly the same as 60Hz.

This is also very useful for R77 and in general people running Stella on a TV, since some TVs may not even offer 60Hz, only 59Hz.

**Comment by thrust26** [comment]
I just looked into changing the resolution too and found that this would require some major code restructuring. The whole code is not prepared for changed resolutions, everything is calculated at the start. When we allow changing the resolution, everything would have to be calculated on-the-fly. I don't think the effort is worth the minor potential gains.

BTW: I think the current behavior could also lead to problems on multi-monitor system with different resolutions. If we want to do it perfectly right, we would have to switch to on-the-fly calculations there too. If this eventually has been taken care of (see #531), we can come back to this.

**Comment by sa666666** [comment]
The latest code works on my Windows system now.  It uses 59Hz instead of 60Hz.  Thanks for that.

I still need to hook up a 1080p monitor to test 50Hz modes.  I will get back to you on that.

Regarding switching the res, I guess we could do it only for TIA mode.  That way, all we need to worry about are the various menu UI's, not the debugger or ROM launcher.  But that would imply a mode switch every time one switches from emulation to the debugger/launcher.  I agree this is a symptom of a larger problem; of deciding up front on UI sizes vs. sizing on the fly.

**Comment by thrust26** [comment]
Thanks! Unless you find any problems, I suppose the code is done now.

**Comment by DirtyHairy** [comment]
> So IMO we should just try to change the refresh rate. At least for now.

We should also keep in  mind that reducing the resolution will reduce the quality of upscaling, so there's a trade-off there.

**Comment by sa666666** [comment]
I'm refactoring the latest code very slightly, and adding another define for `ADAPTABLE_REFRESH_SUPPORT` rather than tying it to Mac specifically.  This allows more leeway in the future, and better documents what's going on.

Also, I'm removing some of the Mac-specific handling in `FrameBufferSDL2`, since with the latest SDL2 there is no difference in behaviour between other systems (using OpenGL and Direct3D) and Mac (using Metal).  In fact, using the same code on all systems fixes a display bug I'm seeing in OSX.

**Comment by sa666666** [comment]
@thrust26, there is some code in `FrameBufferSDL2` that is commented out with `#ifdef 0`.  Can this now be removed?

**Comment by thrust26** [comment]
That code was meant for changing the display size too. You can remove it.

**Comment by thrust26** [comment]
> > So IMO we should just try to change the refresh rate. At least for now.
> 
> We should also keep in mind that reducing the resolution will reduce the quality of upscaling, so there's a trade-off there.

Yup, But I suppose usually changing the refresh rate is not necessary. And if it is, the resolution should be usually close to the maximum resolution.

**Comment by sa666666** [comment]
OK, I was finally able to test in Windows with a 1080p monitor that exposes 50Hz mode.  Everything seems to be working great!  And whatever the problem was with the sound cutting out when in 'adapt' mode, it is no longer there.  Weird??

Anyway, I guess this can be merged to master whenever you like.

**Comment by thrust26** [comment]
Implemented with 1fb1809
